ctapipe/core/traits.py

```diff
@@ -138,6 +138,65 @@ def has_traits(cls, ignore=("config", "parent")):
     return bool(set(cls.class_trait_names()) - set(ignore))
 
 
+class TelescopeParameterList(list):
```

Comment:
Do we need to inherit from list here? I think this adds many unwanted interfaces, right? Shouldn't we just store the initial values in a member list instance?

Reply:
The previous traitlet was a list, so this PR doesn't change anything about unwanted interfaces. But you might be right that we would prefer defining our own class for this purpose.

Reply:
But on the otherside, with just inheriting from list it works with parsing config files?

Reply:
Probably good as it is, but I'm always a bit hesitant about inheriting from builtins.

Reply:
Sorry, I just noticed this after it was merged - yes it does seem strange. Also, it's not a list but a dict, right? You give it a telescope id and it gives you a value? (the tel_ids are not necessarily contiguous) Perhaps we need to make another PR.

Also, what is the `klass` attribute of TelescopeParameter?

Reply:
It's a list in the sense that what is in the config file is a list of these 3 tuples `(selector, selection, value)`. So probably by inheriting from `list`, everything just works concerning loading / saving config files.

Reply:
```python
class SomeComponent(Component):
   tel_param1 = IntTelescopeParameter(
      default_value=[("type", "*", 10), ("type", "LST*", 100)]
   )

def __init__(self, subarray):
   self.tel_param1.attach_subarray(subarray)

def __call__(self, telid):
   print(self.tel_param1)
   param1 = self.tel_param1[telid]
   print(param1)

```

In this example the first print statement will output `[("type", "*", 10), ("type", "LST*", 100)]`, which is a TelescopeParameterList (subclass of list).
The second print will output `10` (unless its an LST in which case it will output `100`. We have overloaded the `__getitem__` method of a normal list.

Reply:
I think my confusion was that in the original implementation, the traitlet just checks that the user has created a list of "patterns" (triplets), but then the Resolver converted that into  a *dict* of value per telescope id.  But now there are 3 steps: `TelescopeParameter` → `TelescopeParameterList` (replacing the list of tuples used in the original representation of the pattern) → dict (by using __getitem__), is that right?  

So it seems ok, just maybe needs better explanation since it's complicated.

Reply:
Yeah I think the confusing part is that the `__getitem__` of `TelescopeParameterList` operates more like a dict instead of a list, but this was done for conveniance.

Reply:
It's recommended to never subclass `list` in python, so maybe making it an attribute as @MaxNoe said is better.

Reply:
> `TelescopeParameterList` (replacing the list of tuples used in the original representation of the pattern)

`TelescopeParameterList` **IS** the list of tuples.